### ISSUE TITLE
Error after upgrade to v0.6.0-alpha.2: invalid interpolation: cannot add field ...: was already used

### ISSUE THREAD

**Issue by bozaro** [cue-lang/cue#2473]
### What version of CUE are you using (`cue version`)?

v0.6.0-alpha.2
<pre>
$ cue version

cue version v0.0.0-20230628071453-b6a32da097fc

go version go1.20.5
      -buildmode exe
       -compiler gc
     CGO_ENABLED 1
          GOARCH amd64
            GOOS linux
         GOAMD64 v1
             vcs git
    vcs.revision b6a32da097fce5f9c960ad0a02da51efb18de2fd
        vcs.time 2023-06-28T07:14:53Z
    vcs.modified true
</pre>

### Does this issue reproduce with the latest stable release?
No

### What did you do?
I try to upgrade from v0.6.0-alpha.1 to v0.6.0-alpha.2.

`cue export` for file (https://cuelang.org/play/?id=b28n-fE2zNC#cue@export@cue):
```bash
cd $(mktemp -d)
(
set -e
cat <<EOD > upgrade_issue.cue
package main

#ClusterInfo: {
	env_type: string
}

#ClusterInfoByEnvName: #ClusterInfo & {
	#env:     "dev" | "stage" | "production"
	env_type: #env
}

#ProductionEndpoints: {
	#cluster_info: #ClusterInfo

	example_host: "\(#cluster_info.env_type).example.net"
}

endpoints: {
	#cluster_info: #ClusterInfoByEnvName & {
		#env: "production"
	}

	if #cluster_info.env_type == "production" {
		#ProductionEndpoints
	}
}
EOD
cue export upgrade_issue.cue
)
```

### What did you expect to see?
With v0.6.0-alpha.1 and v0.5.1 I got result like:
```
endpoints: {
	example_host: "production.example.net"
}
```

### What did you see instead?
With v0.6.0-alpha.2 I got error like:
```
endpoints: invalid interpolation: cannot add field env_type: was already used:
    ./upgrade_issue.cue:15:16
    ./upgrade_issue.cue:4:2
```

**Comment by mvdan** [comment]
Reduced the `cue export` input to:

```
c: {
	env:  "prod"
	env:  "dev" | "prod"
	env2: env
}

if c.env2 == "prod" {
	{// removing this block fixes master
		c: env2: string
		example: c.env2
	}
}
```

This snippet still worked with v0.5.0, and fails with `cannot add field env2: was already used` on master, so I've bisected it to https://review.gerrithub.io/c/cue-lang/cue/+/555953.

There's also this snippet, which fails on both v0.5.0 and master:

```
env:  "prod"
env:  "dev" | "prod"
env2: env

if env2 == "prod" {
	{// removing this block fixes v0.5.0 and master
		env2:    string
		example: env2
	}
}
```

**Comment by mvdan** [comment]
@mpvl correctly points out that this is likely a duplicate of #2337, which is already milestoned for v0.6.0. The reproducers are quite different, and yours bisects to a recent change, but note how the change was https://review.gerrithub.io/c/cue-lang/cue/+/555953, which simply surfaces errors which the evaluator used to incorrectly ignore entirely. So the underlying bug was already present, and it's most likely the same as #2337.

I'm going to close this as a duplicate, and add the reproducers from here to that thread, to make sure they are included as regression tests with the fix as well. Thank you again for filing this bug!
